Match requests behavior with proxies dict, support "all" scheme

Velocidensity opened the pull request:
Previously, curl_cffi would crash if given a dict with no proxy set for a given scheme. Now, it will use the value per scheme if set, otherwise fallback to all (if set). Also allows proxies to be set per scheme+hostname, similar to requests.

Reference: [psf/requests: requests/utils.py#L833-L856](https://github.com/psf/requests/blob/769bc3ae5cc8cb42242e4505b310dd10889cb54c/src/requests/utils.py#L833-L856)

yifeikong commented on curl_cffi/requests/session.py:
```diff
@@ -335,21 +335,27 @@ def _set_curl_options(
-                    if not proxies["https"].startswith("socks"):
-                        c.setopt(CurlOpt.HTTPPROXYTUNNEL, 1)
+            parts = urlparse(url)
+            proxy = proxies.get(parts, proxies.get('all'))
```
shouldn't this be `parts.scheme`? typo?

Velocidensity replied:
Yes, thanks, corrected.

yifeikong commented:
Please add some unittests to the parts you would like to update.

Velocidensity commented:
Added. I couldn't think of a better way to test proxies than mocking one, so this adds a new dependency. Let me know if that approach is not to your liking. Also, squashed previous commits into one to avoid clogging the history.

T-256 commented on curl_cffi/requests/session.py:
```diff
@@ -335,21 +335,27 @@ def _set_curl_options(
+                ) or proxy
+
+            if proxy is not None:
+                if parts.scheme == "https" and proxy.startswith("https://"):
+                    raise RequestsError(
+                        "You are using http proxy WRONG, the prefix should be 'http://' not 'https://',"
+                        "see: https://github.com/yifeikong/curl_cffi/issues/6"
+                    )
```
In fact, curl supports HTTPS proxies [since v7.52](https://daniel.haxx.se/blog/2016/11/26/https-proxy-with-curl/).

Could it possible to support it in `curl-cffi`?
Also, recently `httpcore` (core of `httpx`): https://github.com/encode/httpcore/pull/745

yifeikong replied:
Yes, I'm aware of that. However, in real worlds, no one uses TLS-in-TLS proxies, they just mistakenly think that http proxy for https urls should start with `https://`.

T-256 replied:
> they just mistakenly think that http proxy for https urls should start with `https://`.

Then we could show Error only when we are sure it used mistakenly.

Instead of raise error before making requests:
```py
(CurlError("Failed to perform, ErrCode: 35, Reason: 'error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER'"),)
```

We could wrap CurlError if proxy scheme is HTTPS, then show additional warning:
```py
     warning.warn(
        "You are using httpS proxy, we noticed proxy-server doesn't support HTTPS,"
        " try replace proxy protocol 'https://' with 'http://' and check if it's helping."
     )
```

Velocidensity replied:
NordVPN proxies use https, I use them sometimes.

T-256 reviewed (approved):
LGTM, Thank you!
